pinot-controller/src/main/java/org/apache/pinot/controller/util/ServerSegmentMetadataReader.java

```diff
@@ -47,6 +51,73 @@ public ServerSegmentMetadataReader(Executor executor, HttpConnectionManager conn
   }
 
+  /**
+   * This method is called when the API request is to fetch aggregated segment metadata for all segments of the table.
+   * This method makes a MultiGet call to all servers that host their respective segments and gets the results.
+   * This method accept a list of column names as filter, and will return column metadata for the column in the
+   * list.
+   * TODO Some performance improvement ideas to explore:
+   * - If table has replica groups, only send requests to one replica group.
+   * - If table does not have replica groups, send requests to a minimal set of servers hosting all segments of the
+   *   table.
+   */
+  public AggregateTableMetadataInfo getAggregatedTableMetadataFromServer(String tableNameWithType,
+      BiMap<String, String> serverEndPoints, List<String> columns, int numReplica, int timeoutMs) {
+    int numServers = serverEndPoints.size();
+    LOGGER.info("Reading aggregated segment metadata from {} servers for table: {} with timeout: {}ms", numServers,
+        tableNameWithType, timeoutMs);
+
+    List<String> serverUrls = new ArrayList<>(numServers);
+    BiMap<String, String> endpointsToServers = serverEndPoints.inverse();
+    for (String endpoint : endpointsToServers.keySet()) {
+      String serverUrl = generateAggregateSegmentMetadataServerURL(tableNameWithType, columns, endpoint);
+      serverUrls.add(serverUrl);
+    }
+
+    // Helper service to run a http get call to the server
+    CompletionServiceHelper completionServiceHelper =
+        new CompletionServiceHelper(_executor, _connectionManager, endpointsToServers);
+    CompletionServiceHelper.CompletionServiceResponse serviceResponse =
+        completionServiceHelper.doMultiGetRequest(serverUrls, tableNameWithType, false, timeoutMs);
+
+    AggregateTableMetadataInfo aggregateTableMetadataInfo = new AggregateTableMetadataInfo();
+    int totalNumSegments = 0;
+    int failedParses = 0;
+    for (Map.Entry<String, String> streamResponse : serviceResponse._httpResponses.entrySet()) {
+      try {
+        TableMetadataInfo tableMetadataInfo =
+            JsonUtils.stringToObject(streamResponse.getValue(), TableMetadataInfo.class);
+        aggregateTableMetadataInfo.diskSizeInBytes += tableMetadataInfo.diskSizeInBytes;
+        aggregateTableMetadataInfo.numRows += tableMetadataInfo.numRows;
+        totalNumSegments += tableMetadataInfo.numSegments;
+        tableMetadataInfo.columnLengthMap
+            .forEach((k, v) -> aggregateTableMetadataInfo.columnAvgLengthMap.merge(k, (double) v, Double::sum));
+        tableMetadataInfo.columnCardinalityMap
+            .forEach((k, v) -> aggregateTableMetadataInfo.columnAvgCardinalityMap.merge(k, (double) v, Double::sum));
+      } catch (IOException e) {
+        failedParses++;
+        LOGGER.error("Unable to parse server {} response due to an error: ", streamResponse.getKey(), e);
+      }
+    }
+
+    final int finalTotalNumSegments = totalNumSegments;
```

Comment:
(nit) we don't use final in our current coding guidelines so suggest removing it